src/manifold/src/impl.cpp

```diff
@@ -824,37 +815,20 @@ void Manifold::Impl::CalculateNormals() {
  * instances of these meshes.
  */
 void Manifold::Impl::IncrementMeshIDs() {
-  const int numTri = NumTri();
-  const auto policy = autoPolicy(numTri);
   HashTable<uint32_t> meshIDold2new(meshRelation_.meshIDtransform.size() * 2);
-
-  while (1) {
-    for_each_n(policy, meshRelation_.triRef.begin(), numTri,
-               MarkMeshID({meshIDold2new.D()}));
-    if (!meshIDold2new.Full()) break;
-    meshIDold2new = HashTable<uint32_t>(meshIDold2new.Size() * 2);
-  }
-  inclusive_scan(autoPolicy(meshIDold2new.Size()),
-                 meshIDold2new.GetValueStore().begin(),
-                 meshIDold2new.GetValueStore().end(),
-                 meshIDold2new.GetValueStore().begin());
-  const int numMeshIDs = meshIDold2new.GetValueStore().back();
-  const int meshIDstart = ReserveIDs(numMeshIDs);
-  // We do start - 1 because the inclusive scan makes our first index 1
-  // instead of 0.
-  for_each_n(policy, meshRelation_.triRef.begin(), numTri,
-             UpdateMeshID({meshIDold2new.D(), meshIDstart}));
   // Update keys of the transform map
   std::map<int, Relation> oldTransforms;
   std::swap(meshRelation_.meshIDtransform, oldTransforms);
-  const int tableSize = meshIDold2new.Size();
-  for (int i = 0; i < tableSize; ++i) {
-    const auto oldID = meshIDold2new.D().KeyAt(i);
-    if (oldID != HashTable<uint32_t>::Open()) {
-      meshRelation_.meshIDtransform[meshIDold2new.D().At(i) + meshIDstart] =
-          oldTransforms[oldID];
-    }
+  const int numMeshIDs = oldTransforms.size();
+  int nextMeshID = ReserveIDs(numMeshIDs);
+  for (const auto& pair : oldTransforms) {
+    meshIDold2new.D().Insert(pair.first, nextMeshID);
+    meshRelation_.meshIDtransform[nextMeshID++] = pair.second;
```

Comment:
Interesting, will have a look at it.

Reply:
OK, I think I saw an issue here. We need to perform atomic load as well, as normal load together with atomic write is undefined behavior. This is not an issue with CUDA because all aligned loads on CUDA are atomic (https://stackoverflow.com/questions/36101775/cuda-is-there-an-atomicread), but this will be problematic on the CPU.

Reply:
So the `Full`, `At`, `KeyAt`, `[]` methods all require atomic load.

Reply:
I think I can send a PR later today to test this

Reply:
Awesome, thanks!